compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -675,4 +679,485 @@ object Erasure {
+    *    type parameter of intersectionDominator(List(El1, El2, ...)).           <--- @PP: not yet in spec.
+    *  - Otherwise, the list is reduced to a subsequence containing only types
+    *    which are not subtypes of other listed types (the span.)
+    *  - If the span is empty, the dominator is Object.
+    *  - If the span contains a class Tc which is not a trait and which is
+    *    not Object, the dominator is Tc.                                        <--- @PP: "which is not Object" not in spec.
+    *  - Otherwise, the dominator is the first element of the span.
+    */
+  private def intersectionDominator(parents: List[Type])(implicit ctx: Context): Type = {
+    if (parents.isEmpty) defn.ObjectType
+    else {
+      val psyms = parents map (_.typeSymbol)
+      if (psyms contains defn.ArrayClass) {
+        // treat arrays specially
+        defn.ArrayType.appliedTo(intersectionDominator(parents.filter(_.typeSymbol == defn.ArrayClass).map(t => t.typeParams.head.paramInfo)))
+      } else {
+        // implement new spec for erasure of refined types.
+        def isUnshadowed(psym: Symbol) =
+          !(psyms exists (qsym => (psym ne qsym) && (qsym isSubClass psym)))
+        val cs = parents.iterator.filter { p => // isUnshadowed is a bit expensive, so try classes first
+          val psym = p.typeSymbol
+          psym.ensureCompleted()
+          psym.isClass && !psym.is(Flags.Trait) && isUnshadowed(psym)
+        }
+        (if (cs.hasNext) cs else parents.iterator.filter(p => isUnshadowed(p.typeSymbol))).next()
+      }
+    }
+  }
+
+  /* Drop redundant types (ones which are implemented by some other parent) from the immediate parents.
+   * This is important on Android because there is otherwise an interface explosion.
+   */
+  private def minimizeParents(cls: Symbol, parents: List[Type])(implicit ctx: Context): List[Type] = if (parents.isEmpty) parents else {
+    // val requiredDirect: Symbol => Boolean = requiredDirectInterfaces.getOrElse(cls, Set.empty)
+    var rest   = parents.tail
+    var leaves = collection.mutable.ListBuffer.empty[Type] += parents.head
+    while (rest.nonEmpty) {
+      val candidate = rest.head
+      val candidateSym = candidate.typeSymbol
+      // val required = requiredDirect(candidateSym) || !leaves.exists(t => t.typeSymbol isSubClass candidateSym)
+      val required = !leaves.exists(t => t.typeSymbol.isSubClass(candidateSym))
+      if (required) {
+        leaves = leaves filter { t =>
+          val ts = t.typeSymbol
+          !(ts.is(Flags.Trait) || ts.is(Flags.PureInterface)) || !candidateSym.isSubClass(ts)
+          // requiredDirect(ts) || !ts.isTraitOrInterface || !candidateSym.isSubClass(ts)
+        }
+        leaves += candidate
+      }
+      rest = rest.tail
+    }
+    leaves.toList
+  }
+
+  private def hiBounds(bounds: TypeBounds)(implicit ctx: Context): List[Type] = bounds.hi.widenDealias match {
+    case AndType(tp1, tp2) => tp1 :: tp2 :: Nil
```

Comment:
What happens with nested AndTypes? Don't you need a recursive call here?

Reply:
Interesting. I thought at first that it was not necessary, because all the parents would be picked up in `jsig` and the signature would be correct, but further investigation showed that I was wrong, and it's actually a bug in scalac:

Considering the following definitions:
```scala
class C1 { def c1 = 0 }
trait T1 { def t1: Int }
trait T2 { def t2: Int }
class Foo[T <: (C1 with T1) with T2] { // Note the parentheses
  def foo(t: T): Int = t.c1 + t.t1 + t.t2
}
```

The signature of `Foo` that scalac generates is: `<T:LC1;:LT2;>Ljava/lang/Object` There's no mention of `T1`! This means that javac won't complain if someone passes an instance of `C1 with T2` to `def foo`, resulting in a runtime crash.

With the recursive call that you suggest, dotty generates the expected signature: `<T:LC1;:LT1;:LT2;>Ljava/lang/Object;`

Reply:
/cc @retronym

Reply:
Thanks, noted as https://github.com/scala/bug/issues/10543

Reply:
Hrm, looks like another instance of too aggressively replacing normalize (which flattens refined types) with dealias in
```
  private def hiBounds(bounds: TypeBounds): List[Type] = bounds.hi.dealiasWiden match {
    case RefinedType(parents, _) => parents map (_.dealiasWiden)
    case tp                      => tp :: Nil
  }
```